bofire/strategies/predictives/acqf_optimization_utils/GA_utils.py

```diff
@@ -0,0 +1,720 @@
+        self.input_preprocessing_specs = input_preprocessing_specs
+        self.q = q
+
+        for key in domain.inputs.get_keys():
+            spec_ = input_preprocessing_specs.get(key)
+            input_ref = domain.inputs.get_by_key(key)
+
+            if isinstance(input_ref, ContinuousInput) and spec_ is None:
+                # simple case: non-transformed continuous variable
+                bounds_org = np.array(
+                    domain.inputs.get_by_key(key).get_bounds(spec_)
+                ).reshape(-1)
+                self.vars[key] = pymoo_variable.Real(
+                    bounds=bounds_org
+                )  # default variable
+
+            elif (
+                isinstance(input_ref, CategoricalDescriptorInput)
+                and spec_ == CategoricalEncodingEnum.DESCRIPTOR
+            ):
+                # categorical descriptor
+                self.vars[key] = pymoo_variable.Choice(
+                    options=domain.inputs.get_by_key(key).get_allowed_categories(),
+                )
+
+            elif (
+                isinstance(input_ref, CategoricalInput)
+                and spec_ == CategoricalEncodingEnum.ONE_HOT
+            ):
+                # one-hot encoded categorical input
+                self.vars[key] = pymoo_variable.Choice(
+                    options=domain.inputs.get_by_key(key).get_allowed_categories(),
+                )
+
+            elif isinstance(input_ref, DiscreteInput) and spec_ is None:
+                # numerical discrete input
+                conversion = dict(enumerate(input_ref.values))
+                self.vars[key] = pymoo_variable.Integer(bounds=[0, len(conversion) - 1])
+                self.pymoo_conversion[key] = conversion
+
+            else:
+                raise NotImplementedError(
+                    f"Input {input_ref.key}: Input type {type(input_ref)} with preprocessing spec {spec_} not supported"
+                )
+
+    @property
+    def pymoo_conversion_inverse(self) -> Dict[str, dict]:
+        return {
+            factor: {value: key for (key, value) in conv_dict.items()}
+            for (factor, conv_dict) in self.pymoo_conversion.items()
+        }
+
+    def pymoo_vars(self) -> Dict[str, pymoo_variable.Variable]:
+        """return the variables in the format required by pymoo. Includes repeats for q-points.
+
+        Returns:
+            Dict[str, pymoo_variable.Variable]: dictionary, to be used in pymoo.Problem __init__. The keys
+
+        """
+        vars = {}
+        for qi in range(self.q):
+            vars = {**vars, **{f"{key}_q{qi}": var for key, var in self.vars.items()}}
+        return vars
+
+    def _pymoo_specific_transform(self, experiments: pd.DataFrame) -> pd.DataFrame:
+        """handles the non-domain encodings (e.g. pymoo-type 'Integer' to discrete"""
+        for key, conversion in self.pymoo_conversion.items():
+            experiments[key] = [conversion[i] for i in experiments[key]]
```

Comment:
Can you use pandas.DataFrame.replace here to get rid of the the second loop (comprehension)? https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.replace.html

Reply:
See the comment also below, it could be that we could completly drop it.